From c119fa15a860f2ff6713936f0ad97e922c1cd65f Mon Sep 17 00:00:00 2001
From: Noam Camus <ncamus@habana.ai>
Date: Mon, 15 Apr 2019 15:41:55 +0300
Subject: [PATCH] [SW-0] pci: wait for boot manager to reach idle state

Boot Manager after wakening CPU from primary image starts executing PCIe
image (depends on pcie_enable strap).
Boot Manager execute straight from SPI flash.
Waiting for it to reach idle state insures there is no contention on SPI
bus which will result on boot hang due to curropted data read from
flash.

Change-Id: I1b48e1a6e26249e80a447f423cd02081ba0e7116
---

diff --git a/common/preboot.c b/common/preboot.c
index dfbd364..3325291 100644
--- a/common/preboot.c
+++ b/common/preboot.c
@@ -140,6 +140,16 @@
 static inline void preboot_version(void) {}
 #endif	/* TARGET_IS_TESTER */
 
+/*
+ * Wait for Boot Manager to complete PCIe image execution
+ * and return to idle state, so it won't initiate SPI access anymore
+ */
+static void preboot_sync_btl(void)
+{
+	while (regmap_readl(mmPSOC_GLOBAL_CONF_BTM_FSM))
+		;
+}
+
 void preboot_main(void)
 {
 	void (*spl_start)(void) = (void *)SPL_TEXT_BASE;
@@ -149,6 +159,7 @@
 	preboot_version();
 	preboot_pcie_init();
 	preboot_delay();
+	preboot_sync_btl();
 	hl_pll_init();
 	preboot_copy_spl();
 
